net/modemmanager/files/modemmanager.proto

```diff
@@ -0,0 +1,413 @@
+	proto_config_add_boolean lowpower
+}
+
+proto_modemmanager_setup() {
+	local interface="$1"
+
+	local modempath modemstatus bearercount bearerpath connectargs bearerstatus beareriface
+	local operatorname operatorid registration accesstech signalquality
+
+	local device apn username password pincode iptype metric
+	
+	local address prefix gateway mtu dns1 dns2
+
+	json_get_vars device apn username password pincode iptype metric
+
+	# validate sysfs path given in config
+	[ -n "${device}" ] || {
+		echo "No device specified"
+		proto_notify_error "${interface}" NO_DEVICE
+		proto_set_available "${interface}" 0
+		return 1
+	}
+	[ -e "${device}" ] || {
+		echo "Device not found in sysfs"
+		proto_set_available "${interface}" 0
+		return 1
+	}
+
+	# validate that ModemManager is handling the modem at the sysfs path
+	modemstatus=$(mmcli --modem="${device}" --output-keyvalue)
+	modempath=$(modemmanager_get_field "${modemstatus}" "modem.dbus-path")
+	[ -n "${modempath}" ] || {
+		echo "Device not managed by ModemManager"
+		proto_notify_error "${interface}" DEVICE_NOT_MANAGED
+		proto_set_available "${interface}" 0
+		return 1
+	}
+	echo "modem available at ${modempath}"
+
+	# always cleanup before attempting a new connection, just in case
+	modemmanager_cleanup_connection "${modemstatus}"
+
+	# setup connect args; APN mandatory (even if it may be empty)
+	echo "starting connection with apn '${apn}'..."
+	connectargs="apn=${apn}${username:+,user=${username}}${password:+,password=${password}}${pincode:+,pin=${pincode}}${iptype:+,ip-type=${iptype}}"
+	mmcli --modem="${device}" --timeout 120 --simple-connect="${connectargs}" || {
+		proto_notify_error "${interface}" CONNECT_FAILED
+		proto_block_restart "${interface}"
+		return 1
+	}
+
+	# log additional useful information
+	modemstatus=$(mmcli --modem="${device}" --output-keyvalue)
+	operatorname=$(modemmanager_get_field "${modemstatus}" "modem.3gpp.operator-name")
+	[ -n "${operatorname}" ] && echo "network operator name: ${operatorname}"
+	operatorid=$(modemmanager_get_field "${modemstatus}" "modem.3gpp.operator-code")
+	[ -n "${operatorid}" ] && echo "network operator MCCMNC: ${operatorid}"
+	registration=$(modemmanager_get_field "${modemstatus}" "modem.3gpp.registration-state")
+	[ -n "${registration}" ] && echo "registration type: ${registration}"
+	accesstech=$(modemmanager_get_multivalue_field "${modemstatus}" "modem.generic.access-technologies")
+	[ -n "${accesstech}" ] && echo "access technology: ${accesstech}"
+	signalquality=$(modemmanager_get_field "${modemstatus}" "modem.generic.signal-quality.value")
+	[ -n "${signalquality}" ] && echo "signal quality: ${signalquality}%"
+
+	# we won't like it if there are more than one bearers, as that would mean the
+	# user manually created them, and that's unsupported by this proto
+	bearercount=$(modemmanager_get_field "${modemstatus}" "modem.generic.bearers.length")
+	[ -n "${bearercount}" ] && [ "$bearercount" -eq 1 ] || {
```

Comment:
This throws https://github.com/koalaman/shellcheck/wiki/SC2015 . Looks legitimate.

Reply:
This looks to me like `if p and q then` r, not if `p then q else r`

Reply:
Yes, the purpose of the check in the code is "if (p and q); then r". That is correct, isn't it?